File: src/core/soul-mesh/SoulHybridChannel.ts
/**
 * Hybrid channel contract for the Soul GPU fabric.
 * A logical IN/OUT port can use different physical transports without changing
 * message identity or semantics. Transport negotiation is deliberately kept
 * separate from nucleus capability ownership.
 */
import type { SoulMeshMessage, SoulNucleus } from '../mesh/SoulMeshProtocol';

export type SoulHybridTransportKind =
  | 'WEBVIEW_BRIDGE'
  | 'LOOPBACK_HTTP'
  | 'HTTP'
  | 'REALTIME'
  | 'IN_PROCESS';

export type SoulMeshDirection = 'in' | 'out';
export type SoulMeshSlot = 1 | 2 | 3 | 4 | 5;

export interface SoulHybridChannel {
  id: `${SoulNucleus}->${SoulNucleus}:${SoulMeshDirection}:${SoulMeshSlot}`;
  source: SoulNucleus;
  target: SoulNucleus;
  direction: SoulMeshDirection;
  slot: SoulMeshSlot;
  transports: readonly SoulHybridTransportKind[];
}

export interface SoulHybridChannelAdapter {
  readonly kind: SoulHybridTransportKind;
  isAvailable(channel: SoulHybridChannel): boolean | Promise<boolean>;
  send(message: SoulMeshMessage): Promise<void>;
}

export const SOUL_HYBRID_TRANSPORT_PRIORITY: readonly SoulHybridTransportKind[] = [
  'IN_PROCESS',
  'WEBVIEW_BRIDGE',
  'LOOPBACK_HTTP',
  'REALTIME',
  'HTTP',
];

export function channelId(source: SoulNucleus, target: SoulNucleus, direction: SoulMeshDirection, slot: SoulMeshSlot): SoulHybridChannel['id'] {
  return `${source}->${target}:${direction}:${slot}`;
}

export function createHybridChannel(source: SoulNucleus, target: SoulNucleus, slot: SoulMeshSlot, direction: SoulMeshDirection): SoulHybridChannel {
  if (source === target) throw new Error('SELF_CHANNEL_NOT_ALLOWED');
  return {
    id: channelId(source, target, direction, slot),
    source,
    target,
    direction,
    slot,
    transports: SOUL_HYBRID_TRANSPORT_PRIORITY,
  };
}

File: soul-sentinel/app/src/main/java/com/divibisoul/soul/SoulHybridChannel.kt
package com.divibisoul.soul

/**
 * Logical channel for the hybrid GPU fabric. A port is stable even when its
 * physical transport changes between in-process, WebView, loopback or network.
 */
data class SoulHybridChannel(
    val source: String,
    val target: String,
    val direction: Direction,
    val slot: Int,
    val transports: List<Transport> = listOf(
        Transport.IN_PROCESS,
        Transport.WEBVIEW_BRIDGE,
        Transport.LOOPBACK_HTTP,
        Transport.REALTIME,
        Transport.HTTP,
    ),
) {
    init {
        require(source != target) { "SELF_CHANNEL_NOT_ALLOWED" }
        require(slot in 1..5) { "INVALID_CHANNEL_SLOT" }
    }

    val id: String get() = "$source->$target:${direction.name.lowercase()}:$slot"

    enum class Direction { IN, OUT }
    enum class Transport { WEBVIEW_BRIDGE, LOOPBACK_HTTP, HTTP, REALTIME, IN_PROCESS }
}

object SoulHybridTopology {
    val nuclei = listOf("N01", "N02", "N03", "N04", "N05", "N06")

    fun channels(owner: String): List<SoulHybridChannel> {
        require(owner in nuclei) { "Unknown nucleus: $owner" }
        val peers = nuclei.filterNot { it == owner }
        return peers.mapIndexed { index, peer ->
            val slot = index + 1
            listOf(
                SoulHybridChannel(owner, peer, SoulHybridChannel.Direction.OUT, slot),
                SoulHybridChannel(peer, owner, SoulHybridChannel.Direction.IN, slot),
            )
        }.flatten()
    }

    fun allChannels(): List<SoulHybridChannel> = nuclei.flatMap(::channels)
}
